source/language/openpulse.rst

```diff
@@ -0,0 +1,316 @@
+   gaussian_square(length:l, complex[float[32]]:amp, length:center, length:square_width, length:sigma)
+
+   // amp is pulse amplitude at center
+   // center is the mean of pulse
+   // sigma is the standard deviation of pulse
+   // beta is the Y correction amplitude, see the DRAG paper
+   drag(length:l, complex[float[32]]:amp, length:center, length:sigma, float[32]:beta)
+
+   // Define a boxcar (aka. constant) pulse of length l
+   boxcar(l:length)
+
+Play instruction
+~~~~~~~~~~~~~~~~
+
+Pulses are scheduled using the ``play`` instruction. Play instructions have two
+required parameters:
+
+- the channel on which to play the pulse
+- a value of type ``pulse`` representing the pulse envelope
+
+Optionally the ``play`` instruction can also take a third parameter: the frame
+to use for the pulse. If the frame is not specified then the corresponding frame
+for the channel will be used instead, as if ``frameof`` was called for that
+channel.
+
+.. code-block:: none
+
+   // Play a 3 sample pulse on qubit 0
+   play transmit(%0), [1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j];
+
+   // Play a gaussian on qubit 0 using qubit 1's frame
+   frame f1 = frameof(transmit(%1));
+   play transmit(%0), gaussian(...), f1;
+
+Capture Instruction
+-------------------
+
+Acquisition is scheduled by a ``capture`` instruction. This is a special
+``kernel`` function which is specified by a hardware vendor. The measurement
+process is difficult to describe generically due to the wide variety of
+hardware and measurement methods.
+
+The only required parameter is an ``channel``, ie. a channel returned from
+calling the ``receive`` function.
+
+The following are possible parameters that might be included:
+
+- A "duration" of type ``length``, if it cannot be inferred from other parameters
+- A "filter", which is dot product-ed with the measured IQ the distill the
+  result into a single IQ value
+- A "tag", which could be used to identify which branch of an if statement was
+  traversed
+
+Again it is up to the hardware vendor to determine the parameters and write a
+kernel definition at the top-level, such as:
+
+.. code-block:: none
+
+   // Minimum requirement
+   kernel capture(channel output) -> complex[32];
+
+   // A capture command with more features
+   kernel capture(channel output, pulse filter) -> complex[32];
+
+The return type of a ``capture`` command varies. It could be a raw trace, ie a
+list of samples taken over a short period of time. It could be some averaged IQ
+value. It could be a classified bit. Or it could even have no return value,
+pushing the results into some buffer which is then accessed outside the program.
+
+Timing
+------
+
+Each channel maintains its own "clock". When a pulse is played the clock for 
+that channel advances by the length of the pulse. The same is true
+for output channels. Pulses on a single channel cannot be played simultaneously,
+although pulses on multiple channels for the same qubit can.
```

Comment:
Is it the channel or the frame the maintains the clock? It seems like the latter would be better, because then I can express things like multiplexed measurement as instructions to separate frames that all share a common channel.

Reply:
Having thought about this more it seems obvious to me now that frames should maintain the clock.